src/main/java/ch/njol/skript/effects/EffToggle.java

```diff
@@ -1,84 +1,181 @@
-@Description("Toggle the state of a block.")
+@Description("Toggle the state of a block or boolean.")
 @Examples({"# use arrows to toggle switches, doors, etc.",
 		"on projectile hit:",
 		"\tprojectile is arrow",
-		"\ttoggle the block at the arrow"})
-@Since("1.4")
+		"\ttoggle the block at the arrow",
+		"",
+		"# With booleans",
+		"toggle gravity of player"
+})
+@Since("1.4, INSERT VERSION (booleans)")
 public class EffToggle extends Effect {
-	
+
+	private enum Action {
+		ACTIVATE, DEACTIVATE, TOGGLE;
+		
+		public boolean apply(boolean current) {
+			return switch(this) {
+				case ACTIVATE -> true;
+				case DEACTIVATE -> false;
+				case TOGGLE -> !current;
+			};
+		}
+	}
+
+	private enum Type {
+		BLOCKS, BOOLEANS, MIXED;
+		
+		/**
+		 * Determines the appropriate Type based on the return type of an expression.
+		 * @param returnType The class representing the return type
+		 * @return The corresponding Type
+		 */
+		public static Type fromClass(Class<?> returnType) {
+			boolean isBlockType = Block.class.isAssignableFrom(returnType);
+			boolean isBooleanType = Boolean.class.isAssignableFrom(returnType);
+			
+			if (isBlockType && !isBooleanType) {
+				return BLOCKS;
+			} else if (isBooleanType && !isBlockType) {
+				return BOOLEANS;
+			} else {
+				return MIXED;
+			}
+		}
+	}
+
+	private static final Patterns<Action> patterns = new Patterns<>(new Object[][]{
+		{"(open|turn on|activate) %blocks%", Action.ACTIVATE},
+		{"(close|turn off|de[-]activate) %blocks%", Action.DEACTIVATE},
+		{"(toggle|switch) [[the] state of] %blocks/booleans%", Action.TOGGLE}
+	});
+
 	static {
-		Skript.registerEffect(EffToggle.class, "(close|turn off|de[-]activate) %blocks%", "(toggle|switch) [[the] state of] %blocks%", "(open|turn on|activate) %blocks%");
+		Skript.registerEffect(EffToggle.class, patterns.getPatterns());
 	}
 
-	@SuppressWarnings("null")
-	private Expression<Block> blocks;
-	private int toggle;
-	
-	@SuppressWarnings({"unchecked", "null"})
+	private Expression<?> togglables;
+	private Action action;
+	private Type type;
+
 	@Override
-	public boolean init(final Expression<?>[] vars, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		blocks = (Expression<Block>) vars[0];
-		toggle = matchedPattern - 1;
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		togglables = expressions[0];
+		action = patterns.getInfo(matchedPattern);
+
+		// Determine expression type using the enum method
+		type = Type.fromClass(togglables.getReturnType());
+		
+		// Validate based on type
+		if (type == Type.BOOLEANS && 
+			!Changer.ChangerUtils.acceptsChange(togglables, Changer.ChangeMode.SET, Boolean.class)) {
+			Skript.error("Cannot toggle '" + togglables + "' as it cannot be set to booleans.");
+			return false;
+		} else if (type == Type.MIXED && !Changer.ChangerUtils.acceptsChange(togglables, Changer.ChangeMode.SET, Object.class)) {
```

Comment:
```suggestion
		} else if (type == Type.MIXED && !Changer.ChangerUtils.acceptsChange(togglables, Changer.ChangeMode.SET, Block.class, Boolean.class)) {
```
I think it'd be better to check whether it accepts Blocks and Booleans rather than the Object supertype